src/backend/opencl/Array.cpp

```diff
@@ -228,30 +228,66 @@ namespace opencl
         Array<T> out =  Array<T>(dims, node);
 
         if (evalFlag()) {
+
             if (node->getHeight() >= (int)getMaxJitSize()) {
                 out.eval();
             } else {
+
                 size_t alloc_bytes, alloc_buffers;
                 size_t lock_bytes, lock_buffers;
 
                 deviceMemoryInfo(&alloc_bytes, &alloc_buffers,
                                  &lock_bytes, &lock_buffers);
 
-                if (lock_bytes > getMaxBytes() ||
-                    lock_buffers > getMaxBuffers()) {
+                bool isBufferLimit =
+                    lock_bytes > getMaxBytes() ||
+                    lock_buffers > getMaxBuffers();
+
+
+                bool isNvidia = getActivePlatform() == AFCL_PLATFORM_NVIDIA;
+                // We eval in the following cases.
+                // 1. Too many bytes are locked up by JIT causing memory pressure.
+                // Too many bytes is assumed to be half of all bytes allocated so far.
+                // 2. Too many buffers in a nonlinear kernel cause param space overflow.
+                // Too many buffers comes out to be about 48 (49 including output).
+                // Too many buffers can occur in a tree of size 24 in the worst case scenario.
+                // This error only happens on nvidia devices.
+                // TODO: Find better solution than the following emperical solution.
+                bool isParamLimit = (isNvidia && node->getHeight() > 24);
```

Comment:
Shouldn't this be based on getMaxJitSize()?

Reply:
No. This change is trying to have different thresholds for Linear and Non Linear kernels. For Linear kernels it is still 100. For non linear it is 24.